librad/src/git/fetch/fetcher.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright © 2019-2020 The Radicle Foundation <hello@radicle.foundation>
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::{collections::BTreeMap, convert::TryFrom, net::SocketAddr};
+
+use git_ext as ext;
+
+use super::{CanFetch, FetchResult, Fetcher, Fetchspecs, RemoteHeads};
+use crate::{
+    git::{
+        p2p::url::GitUrl,
+        storage::{self, Storage},
+    },
+    identities::git,
+    internal::klock::KlockGuard,
+    peer::PeerId,
+};
+
+impl CanFetch for Storage {
+    type Error = storage::Error;
+    type Fetcher<'a> = DefaultFetcher<'a>;
+
+    fn fetcher<'a, Addrs>(
+        &'a self,
+        urn: git::Urn,
+        remote_peer: PeerId,
+        addr_hints: Addrs,
+    ) -> Result<Self::Fetcher<'a>, Self::Error>
+    where
+        Addrs: IntoIterator<Item = SocketAddr>,
+    {
+        Ok(DefaultFetcher::new(self, urn, remote_peer, addr_hints)?)
+    }
+}
+
+/// The default [`Fetcher`], which uses the peer-to-peer network for fetching.
+pub struct DefaultFetcher<'a> {
+    urn: git::Urn,
+    remote_peer: PeerId,
+    remote_heads: RemoteHeads,
+    remote: git2::Remote<'a>,
+    _guard: KlockGuard<'a>,
+}
+
+impl<'a> DefaultFetcher<'a> {
+    #[tracing::instrument(skip(storage, urn, addr_hints), fields(urn = %urn), err)]
+    pub fn new<Addrs>(
+        storage: &'a Storage,
+        urn: git::Urn,
+        remote_peer: PeerId,
+        addr_hints: Addrs,
+    ) -> Result<Self, git2::Error>
+    where
+        Addrs: IntoIterator<Item = SocketAddr>,
+    {
+        // TODO: Revisit once we have busy grafting.
+        //
+        // Tradeoffs:
```

Comment:
There is also the forth possibility of `try_lock`ing here, that is fail-fast a `replicate` if one is already in progress. That's also not ideal, because the later `replicate` could actually be from a different remote peer (which may offer different data).

Reply:
Wouldn't a `try_lock` give some indication to wait and try again?

Reply:
Yes, but that means the problem is somewhere else + needs scrutinising the error variants. 

I think I’ll leave this patch queued until I have basic grafting up to see if this is practical.

Reply:
I'm thinking it's harder to spot a lock not being available and so the remote goes away compared to, we couldn't lock, so we wait and try again. But happy to wait and see :)

Reply:
You were right -- we should almost always abort